File: riveredge-frontend/src/apps/kuaizhizao/services/logistics.ts
import { apiRequest } from '../../../services/api';

const BASE = '/apps/kuaizhizao/logistics';

export type LogisticsCarrier = {
  id: number;
  uuid: string;
  code: string;
  name: string;
  carrier_type: string;
  contact_name?: string;
  contact_phone?: string;
  service_hotline?: string;
  settlement_method?: string;
  supplier_id?: number;
  remark?: string;
  is_enabled: boolean;
};

export type Vehicle = {
  id: number;
  uuid: string;
  plate_number: string;
  vehicle_type?: string;
  load_capacity?: number;
  volume_capacity?: number;
  ownership: string;
  carrier_id?: number;
  status: string;
  remark?: string;
  is_enabled: boolean;
};

export type Driver = {
  id: number;
  uuid: string;
  code: string;
  name: string;
  phone?: string;
  license_number?: string;
  ownership: string;
  carrier_id?: number;
  user_id?: number;
  default_vehicle_id?: number;
  remark?: string;
  is_enabled: boolean;
};

export type FreightOrderSource = {
  id?: number;
  source_type: string;
  source_id: number;
  source_code: string;
  partner_name?: string;
};

export type FreightTrackingEvent = {
  id: number;
  event_type: string;
  event_time: string;
  location?: string;
  remark?: string;
  operator_name?: string;
  lng?: number | null;
  lat?: number | null;
};

export type FreightOrder = {
  id: number;
  uuid: string;
  order_code: string;
  business_direction: string;
  transport_mode: string;
  carrier_id?: number;
  carrier_name?: string;
  vehicle_id?: number;
  vehicle_plate?: string;
  driver_id?: number;
  driver_name?: string;
  driver_phone?: string;
  tracking_number?: string;
  sender_phone?: string;
  recipient_phone?: string;
  query_phone?: string;
  origin_address?: string;
  destination_address?: string;
  origin_lng?: number | null;
  origin_lat?: number | null;
  destination_lng?: number | null;
  destination_lat?: number | null;
  planned_depart_at?: string;
  planned_arrive_at?: string;
  actual_depart_at?: string;
  actual_arrive_at?: string;
  status: string;
  remark?: string;
  created_at?: string;
  updated_at?: string;
  sources?: FreightOrderSource[];
  tracking_events?: FreightTrackingEvent[];
  receipt?: {
    signed_by?: string;
    signed_at?: string;
    receipt_result?: string;
    remark?: string;
  } | null;
};

export type FreightPullCandidate = {
  source_type: string;
  source_id: number;
  source_code: string;
  partner_name: string;
  business_direction: string;
  address?: string;
  tracking_number?: string;
  sender_phone?: string;
  recipient_phone?: string;
  pullable?: boolean;
  blocked_reason?: string | null;
};

export type FreightBillItem = {
  id?: number;
  freight_order_id: number;
  freight_order_code?: string;
  fee_type: string;
  amount: number;
  remark?: string;
};

export type FreightBill = {
  id: number;
  uuid: string;
  bill_code: string;
  carrier_id: number;
  carrier_name: string;
  period_start?: string;
  period_end?: string;
  total_amount: number;
  status: string;
  review_status: string;
  payable_id?: number;
  payable_code?: string;
  remark?: string;
  items?: FreightBillItem[];
};

type ListParams = { skip?: number; limit?: number; keyword?: string };

export type CarrierPresetItem = {
  code: string;
  name: string;
  carrier_type: string;
  service_hotline?: string;
  exists: boolean;
};

export async function listCarriers(params: ListParams = {}) {
  return apiRequest<{ items: LogisticsCarrier[]; total: number }>(`${BASE}/carriers`, { method: 'GET', params });
}

export async function listCarrierPresets() {
  return apiRequest<CarrierPresetItem[]>(`${BASE}/carriers/preset-preview`, { method: 'GET' });
}

export async function loadCarrierPresets(codes?: string[]) {
  return apiRequest<{ created: number; skipped: number; updated?: number; message: string }>(`${BASE}/carriers/load-preset`, {
    method: 'POST',
    data: codes != null ? { codes } : undefined,
  });
}

export async function createCarrier(data: Partial<LogisticsCarrier>) {
  return apiRequest<LogisticsCarrier>(`${BASE}/carriers`, { method: 'POST', data });
}

export async function updateCarrier(id: number, data: Partial<LogisticsCarrier>) {
  return apiRequest<LogisticsCarrier>(`${BASE}/carriers/${id}`, { method: 'PUT', data });
}

export async function deleteCarrier(id: number) {
  return apiRequest<void>(`${BASE}/carriers/${id}`, { method: 'DELETE' });
}

export async function listVehicles(params: ListParams & { ownership?: string } = {}) {
  return apiRequest<{ items: Vehicle[]; total: number }>(`${BASE}/vehicles`, { method: 'GET', params });
}

export async function createVehicle(data: Partial<Vehicle>) {
  return apiRequest<Vehicle>(`${BASE}/vehicles`, { method: 'POST', data });
}

export async function updateVehicle(id: number, data: Partial<Vehicle>) {
  return apiRequest<Vehicle>(`${BASE}/vehicles/${id}`, { method: 'PUT', data });
}

export async function deleteVehicle(id: number) {
  return apiRequest<void>(`${BASE}/vehicles/${id}`, { method: 'DELETE' });
}

export async function listDrivers(params: ListParams & { ownership?: string } = {}) {
  return apiRequest<{ items: Driver[]; total: number }>(`${BASE}/drivers`, { method: 'GET', params });
}

export async function createDriver(data: Partial<Driver>) {
  return apiRequest<Driver>(`${BASE}/drivers`, { method: 'POST', data });
}

export async function updateDriver(id: number, data: Partial<Driver>) {
  return apiRequest<Driver>(`${BASE}/drivers/${id}`, { method: 'PUT', data });
}

export async function deleteDriver(id: number) {
  return apiRequest<void>(`${BASE}/drivers/${id}`, { method: 'DELETE' });
}

export async function listFreightOrders(
  params: ListParams & {
    status?: string;
    status_in?: string;
    business_direction?: string;
    uuid?: string;
  } = {},
) {
  return apiRequest<{ items: FreightOrder[]; total: number }>(`${BASE}/freight-orders`, { method: 'GET', params });
}

export async function lookupFreightOrderBySource(sourceType: string, sourceId: number) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/lookup-by-source`, {
    method: 'GET',
    params: { source_type: sourceType, source_id: sourceId },
  });
}

export type AmapMapPublicConfig = {
  configured: boolean;
  js_key?: string | null;
  security_code?: string | null;
};

export async function getLogisticsMapConfig() {
  return apiRequest<AmapMapPublicConfig>('/core/logistics/map-config', { method: 'GET' });
}

export async function getFreightOrder(id: number) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}`, { method: 'GET' });
}

export async function listFreightPullCandidates(
  params: { business_direction?: string; keyword?: string; skip?: number; limit?: number } = {},
) {
  return apiRequest<{ items: FreightPullCandidate[]; total: number }>(`${BASE}/freight-orders/pull-candidates`, {
    method: 'GET',
    params,
  });
}

export async function createFreightOrder(data: Record<string, unknown>) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders`, { method: 'POST', data });
}

export async function updateFreightOrder(id: number, data: Record<string, unknown>) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}`, { method: 'PUT', data });
}

export async function deleteFreightOrder(id: number) {
  return apiRequest<void>(`${BASE}/freight-orders/${id}`, { method: 'DELETE' });
}

export async function dispatchFreightOrder(id: number) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}/dispatch`, { method: 'POST' });
}

export async function shipFreightOrder(id: number) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}/ship`, { method: 'POST' });
}

export async function markFreightOrderInTransit(id: number) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}/in-transit`, { method: 'POST' });
}

export async function arriveFreightOrder(id: number) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}/arrive`, { method: 'POST' });
}

export async function signFreightOrder(id: number, data: Record<string, unknown>) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}/sign-receipt`, { method: 'POST', data });
}

export async function addFreightTrackingEvent(id: number, data: Record<string, unknown>) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}/tracking-events`, { method: 'POST', data });
}

export async function deleteFreightTrackingEvent(id: number, eventId: number) {
  return apiRequest<FreightOrder>(`${BASE}/freight-orders/${id}/tracking-events/${eventId}/remove`, {
    method: 'POST',
  });
}

export async function listFreightBills(params: ListParams & { review_status?: string } = {}) {
  return apiRequest<{ items: FreightBill[]; total: number }>(`${BASE}/freight-bills`, { method: 'GET', params });
}

export async function createFreightBill(data: Record<string, unknown>) {
  return apiRequest<FreightBill>(`${BASE}/freight-bills`, { method: 'POST', data });
}

export async function deleteFreightBill(id: number) {
  return apiRequest<void>(`${BASE}/freight-bills/${id}`, { method: 'DELETE' });
}

export async function listPendingFreightOrdersForBill(
  params: { carrier_id?: number; keyword?: string; skip?: number; limit?: number } = {},
) {
  return apiRequest<{ items: FreightOrder[]; total: number }>(`${BASE}/freight-bills/pending-freight-orders`, {
    method: 'GET',
    params,
  });
}

export type LogisticsTrackEvent = {
  time?: string;
  status?: string;
  description?: string;
  location?: string;
};

export type LogisticsTrackResult = {
  success: boolean;
  carrier?: string;
  company_code?: string;
  tracking_number?: string;
  status?: string;
  events?: LogisticsTrackEvent[];
};

export async function trackLogistics(carrier: string, trackingNumber: string, phone?: string) {
  return apiRequest<LogisticsTrackResult>('/core/logistics/track', {
    method: 'GET',
    params: {
      carrier,
      tracking_number: trackingNumber,
      ...(phone?.trim() ? { phone: phone.trim() } : {}),
    },
  });
}

File: riveredge-backend/src/apps/kuaizhizao/schemas/logistics.py
"""快制造物流管理 Schema"""

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, Field


class LogisticsCarrierBase(BaseModel):
    code: Optional[str] = None
    name: str
    carrier_type: str = "express"
    contact_name: Optional[str] = None
    contact_phone: Optional[str] = None
    service_hotline: Optional[str] = None
    settlement_method: Optional[str] = None
    supplier_id: Optional[int] = None
    remark: Optional[str] = None
    is_enabled: bool = True


class LogisticsCarrierCreate(LogisticsCarrierBase):
    pass


class LogisticsCarrierUpdate(BaseModel):
    name: Optional[str] = None
    carrier_type: Optional[str] = None
    contact_name: Optional[str] = None
    contact_phone: Optional[str] = None
    service_hotline: Optional[str] = None
    settlement_method: Optional[str] = None
    supplier_id: Optional[int] = None
    remark: Optional[str] = None
    is_enabled: Optional[bool] = None


class LogisticsCarrierResponse(LogisticsCarrierBase):
    id: int
    uuid: str
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class VehicleBase(BaseModel):
    plate_number: str
    vehicle_type: Optional[str] = None
    load_capacity: Optional[Decimal] = None
    volume_capacity: Optional[Decimal] = None
    ownership: str = "internal"
    carrier_id: Optional[int] = None
    status: str = "idle"
    remark: Optional[str] = None
    is_enabled: bool = True


class VehicleCreate(VehicleBase):
    pass


class VehicleUpdate(BaseModel):
    plate_number: Optional[str] = None
    vehicle_type: Optional[str] = None
    load_capacity: Optional[Decimal] = None
    volume_capacity: Optional[Decimal] = None
    ownership: Optional[str] = None
    carrier_id: Optional[int] = None
    status: Optional[str] = None
    remark: Optional[str] = None
    is_enabled: Optional[bool] = None


class VehicleResponse(VehicleBase):
    id: int
    uuid: str
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DriverBase(BaseModel):
    code: Optional[str] = None
    name: str
    phone: Optional[str] = None
    license_number: Optional[str] = None
    ownership: str = "internal"
    carrier_id: Optional[int] = None
    user_id: Optional[int] = None
    default_vehicle_id: Optional[int] = None
    remark: Optional[str] = None
    is_enabled: bool = True


class DriverCreate(DriverBase):
    pass


class DriverUpdate(BaseModel):
    name: Optional[str] = None
    phone: Optional[str] = None
    license_number: Optional[str] = None
    ownership: Optional[str] = None
    carrier_id: Optional[int] = None
    user_id: Optional[int] = None
    default_vehicle_id: Optional[int] = None
    remark: Optional[str] = None
    is_enabled: Optional[bool] = None


class DriverResponse(DriverBase):
    id: int
    uuid: str
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class FreightOrderSourceInput(BaseModel):
    source_type: str
    source_id: int
    source_code: str
    partner_name: Optional[str] = None


class FreightOrderSourceResponse(FreightOrderSourceInput):
    id: int

    class Config:
        from_attributes = True


class FreightOrderCreate(BaseModel):
    business_direction: str
    transport_mode: str = "external_carrier"
    carrier_id: Optional[int] = None
    vehicle_id: Optional[int] = None
    driver_id: Optional[int] = None
    tracking_number: Optional[str] = None
    sender_phone: Optional[str] = None
    recipient_phone: Optional[str] = None
    origin_address: Optional[str] = None
    destination_address: Optional[str] = None
    planned_depart_at: Optional[datetime] = None
    planned_arrive_at: Optional[datetime] = None
    remark: Optional[str] = None
    sources: List[FreightOrderSourceInput] = Field(default_factory=list)


class FreightOrderUpdate(BaseModel):
    transport_mode: Optional[str] = None
    carrier_id: Optional[int] = None
    vehicle_id: Optional[int] = None
    driver_id: Optional[int] = None
    tracking_number: Optional[str] = None
    sender_phone: Optional[str] = None
    recipient_phone: Optional[str] = None
    origin_address: Optional[str] = None
    destination_address: Optional[str] = None
    planned_depart_at: Optional[datetime] = None
    planned_arrive_at: Optional[datetime] = None
    remark: Optional[str] = None


class FreightTrackingEventCreate(BaseModel):
    event_type: str
    event_time: Optional[datetime] = None
    location: Optional[str] = None
    remark: Optional[str] = None


class FreightTrackingEventResponse(FreightTrackingEventCreate):
    id: int
    freight_order_id: int
    lng: Optional[float] = None
    lat: Optional[float] = None
    operator_id: Optional[int] = None
    operator_name: Optional[str] = None

    class Config:
        from_attributes = True


class FreightOrderReceiptCreate(BaseModel):
    signed_by: str
    signed_at: Optional[datetime] = None
    receipt_result: str = "full"
    remark: Optional[str] = None
    attachments: Optional[list] = None


class FreightOrderReceiptResponse(FreightOrderReceiptCreate):
    id: int
    freight_order_id: int

    class Config:
        from_attributes = True


class FreightOrderResponse(BaseModel):
    id: int
    uuid: str
    tenant_id: int
    order_code: str
    business_direction: str
    transport_mode: str
    carrier_id: Optional[int] = None
    carrier_name: Optional[str] = None
    vehicle_id: Optional[int] = None
    vehicle_plate: Optional[str] = None
    driver_id: Optional[int] = None
    driver_name: Optional[str] = None
    driver_phone: Optional[str] = None
    tracking_number: Optional[str] = None
    sender_phone: Optional[str] = None
    recipient_phone: Optional[str] = None
    query_phone: Optional[str] = None
    origin_address: Optional[str] = None
    destination_address: Optional[str] = None
    origin_lng: Optional[float] = None
    origin_lat: Optional[float] = None
    destination_lng: Optional[float] = None
    destination_lat: Optional[float] = None
    planned_depart_at: Optional[datetime] = None
    planned_arrive_at: Optional[datetime] = None
    actual_depart_at: Optional[datetime] = None
    actual_arrive_at: Optional[datetime] = None
    status: str
    remark: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    sources: List[FreightOrderSourceResponse] = Field(default_factory=list)
    tracking_events: List[FreightTrackingEventResponse] = Field(default_factory=list)
    receipt: Optional[FreightOrderReceiptResponse] = None

    class Config:
        from_attributes = True


class FreightOrderListResponse(BaseModel):
    items: List[FreightOrderResponse]
    total: int


class FreightPullCandidate(BaseModel):
    source_type: str
    source_id: int
    source_code: str
    partner_name: str
    business_direction: str
    address: Optional[str] = None
    tracking_number: Optional[str] = None
    sender_phone: Optional[str] = None
    recipient_phone: Optional[str] = None
    pullable: bool = True
    blocked_reason: Optional[str] = None


class FreightPullCandidateListResponse(BaseModel):
    items: List[FreightPullCandidate]
    total: int


class FreightBillItemInput(BaseModel):
    freight_order_id: int
    fee_type: str = "base"
    amount: Decimal
    remark: Optional[str] = None


class FreightBillCreate(BaseModel):
    carrier_id: int
    period_start: Optional[date] = None
    period_end: Optional[date] = None
    remark: Optional[str] = None
    items: List[FreightBillItemInput] = Field(default_factory=list)


class FreightBillUpdate(BaseModel):
    carrier_id: Optional[int] = None
    period_start: Optional[date] = None
    period_end: Optional[date] = None
    remark: Optional[str] = None
    items: Optional[List[FreightBillItemInput]] = None


class FreightBillReject(BaseModel):
    rejection_reason: Optional[str] = None


class FreightBillItemResponse(FreightBillItemInput):
    id: int
    freight_order_code: str
    tracking_number: Optional[str] = None

    class Config:
        from_attributes = True


class FreightBillResponse(BaseModel):
    id: int
    uuid: str
    tenant_id: int
    bill_code: str
    carrier_id: int
    carrier_name: str
    period_start: Optional[date] = None
    period_end: Optional[date] = None
    total_amount: Decimal
    status: str
    review_status: str
    reviewer_id: Optional[int] = None
    reviewer_name: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    payable_id: Optional[int] = None
    payable_code: Optional[str] = None
    remark: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    items: List[FreightBillItemResponse] = Field(default_factory=list)

    class Config:
        from_attributes = True


class FreightBillListResponse(BaseModel):
    items: List[FreightBillResponse]
    total: int


class PaginatedCarrierList(BaseModel):
    items: List[LogisticsCarrierResponse]
    total: int


class CarrierPresetItem(BaseModel):
    code: str
    name: str
    carrier_type: str
    service_hotline: Optional[str] = None
    exists: bool = False


class LoadCarrierPresetRequest(BaseModel):
    codes: Optional[List[str]] = Field(None, description="要加载的预设承运商编码，不传则加载全部")


class LoadCarrierPresetResponse(BaseModel):
    created: int
    skipped: int
    updated: int = 0
    message: str


class PaginatedVehicleList(BaseModel):
    items: List[VehicleResponse]
    total: int


class PaginatedDriverList(BaseModel):
    items: List[DriverResponse]
    total: int
